src/main/scala/chisel3/util/MixedVec.scala

```diff
@@ -0,0 +1,122 @@
+    * @param vals Values to create a MixedVec with and assign
+    * @return MixedVec with given values assigned
+    */
+  def apply[T <: Data](vals: Seq[T]): MixedVec[T] = {
+    // Create a wire of this type.
+    val hetVecWire = Wire(MixedVec(vals.map(_.cloneTypeFull)))
+    // Assign the given vals to this new wire.
+    for ((a, b) <- hetVecWire.zip(vals)) {
+      a := b
+    }
+    hetVecWire
+  }
+}
+
+object MixedVec {
+  /**
+    * Create a MixedVec from that holds the given types.
+    * @param eltsIn Element types. Must be Chisel types.
+    * @return MixedVec with the given types.
+    */
+  def apply[T <: Data](eltsIn: Seq[T]): MixedVec[T] = new MixedVec(eltsIn)
+
+  /**
+    * Create a MixedVec from the type of the given Vec.
+    * For example, given a Vec(2, UInt(8.W)), this creates MixedVec(Seq.fill(2){UInt(8.W)}).
+    * @param vec Vec to use as template
+    * @return MixedVec analogous to the given Vec.
+    */
+  def apply[T <: Data](vec: Vec[T]): MixedVec[T] = {
+    MixedVec(Seq.fill(vec.length)(vec.sample_element))
+  }
+}
+
+/**
+  * A hardware array of elements that can hold values of different types/widths,
+  * unlike Vec which can only hold elements of the same type/width.
+  *
+  * @param eltsIn Element types. Must be Chisel types.
+  *
+  * @example {{{
+  * val v = Wire(MixedVec(Seq(UInt(8.W), UInt(16.W), UInt(32.W))))
+  * v(0) := 100.U(8.W)
+  * v(1) := 10000.U(16.W)
+  * v(2) := 101.U(32.W)
+  * }}}
+  */
+@chiselName
+final class MixedVec[T <: Data](private val eltsIn: Seq[T]) extends Record with collection.IndexedSeq[T] {
+  // We want to create MixedVec only with Chisel types.
+  if (compileOptions.declaredTypeMustBeUnbound) {
+    eltsIn.foreach(e => requireIsChiselType(e))
+  }
+
+  // Clone the inputs so that we have our own references.
+  private val elts: IndexedSeq[T] = eltsIn.map(_.cloneTypeFull).toIndexedSeq
+
+  /**
+    * Statically (elaboration-time) retrieve the element at the given index.
+    * @param index Index with which to retrieve.
+    * @return Retrieved index.
+    */
+  def apply(index: Int): T = elts(index)
+
+  /**
+    * Dynamically (via a mux) retrieve the element at the given index.
+    * This is implemented via a mux with the width of the widest element in this MixedVec.
+    * For example, a MixedVec of type Seq(UInt(4.W), UInt(8.W)) will create an 8-bit mux for this operation.
+    * Note: it is up to the user to process the resultant UInt (e.g. unflatten, etc).
+    *
+    * @param index Index to retrieve. If the index is out of range, it will return the first element.
+    * @return Retrieved index as a UInt with the width of the widest element.
+    */
+  def apply(index: UInt): UInt = {
+    requireIsHardware(index, "index must be hardware")
+
+    if (length < 1) {
+      throw new IndexOutOfBoundsException("Collection is empty")
+    }
+
+    MuxLookup(index, elts.head.asUInt, elts.zipWithIndex.map { case (el, el_index) => el_index.U -> el.asUInt })
+  }
```

Comment:
We discussed this in the meeting, but let's see if we can make this return `T` and dynamically pick the right concrete type with `UInt` being the default for full Heterogeneous types. We should then have a test for a MixedVec of all UInts, all SInts, all (the same) Bundle, and then a mix. That or we can remove this, merge the PR, and do that later

Reply:
Hmm, would you have any suggestions on how to do that? I thought about pattern matching to handle the special cases but we can't match on type parameters (i.e. `T match { ... }` isn't valid).

Originally I tried returning T at one point and the problem occurs, as you mentioned in a previous comment, when heterogeneous types get mixed. It would be okay if I could detect if `T` isn't all UInt/SInt/same bundle, but I don't see an obvious way to do that.

Reply:
You can do it with ClassTags: https://www.scala-lang.org/api/2.12.3/scala/reflect/ClassTag.html

I'm not 100% on how to do it but that's the basic approach.

I think it might be worth removing the dynamic index for now and merging MixedVec without it since it's useful for IOs and dynamic indexing can be shimmed with `VecInit(myMixedVec)(idx)` (I *think*).

Reply:
One concern here is that if the MixedVec is composed of 2 different bundles, then their common supertype would be `Bundle`, and I think we get a `ClassCastException` if we try to cast a generic a generic UInt to it.

Perhaps we could keep both a UInt-returning version as well as a T-returning version (possibly Option[T]-returning) with different names and documentation? It seems unlikely that we can cover all cases by returning just T, and in those cases it might still be useful to have a mux function, I think.

Reply:
I'm also struggling to come up with a good and clean solution - perhaps that indicates a deeper problem? Returning a concrete supertype (like Bundle) that carries data / bits would require a major overhaul of behavior, and returning UInt seems special-case and unclean. It might could also be unintuitive ("wait, how am I getting a UInt when this vec of full of Bundle types?!")

I'm trying to think of what the user wants when dynamically indexing such a vec, would using a regular vec with tagged unions (which don't exist yet but are a requested feature) be a cleaner solution?

Anyways, I'm also leaning towards removing this until we can figure out a clean API. It's easier to add features than to remove them, especially if there are alternative (though potentially clunkier) pathways available currently.

Reply:
Alright, done.